### ISSUE TITLE
Article assumes repository

### ISSUE THREAD

**Issue by vanillajonathan** [dotnet/AspNetCore.Docs#15809]
The article assumes that there is a repository. This despite that the [ASP.NET Core MVC with EF Core - tutorial series](https://docs.microsoft.com/en-us/aspnet/core/data/ef-mvc/crud) use a `DbContext` in the controllers.

The API documentation for [`DbContext`](https://docs.microsoft.com/en-us/dotnet/api/microsoft.entityframeworkcore.dbcontext) mentions that it implements the repository pattern.
> DbContext is a combination of the Unit Of Work and Repository patterns.

Some argue that putting a repository over a DbContext is redundant since it already is a repository.

This articles assumes a repository is used without even ~meaning~ mentioning the DbContext. Maybe the article should mention that this article assumes a repository, or whether it recommends the use of a repository class, or whether controller testing is possible when directly using the DbContext, and if so, then what is the recommended approach. Such as mocking or using the in-memory database (is it still unit testing then, or is it a integration test?).

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: b5cce372-24e2-3489-cb19-c85d934e5d3b
* Version Independent ID: 9d12b19f-a94e-4df4-357b-3f3b93696eb6
* Content: [Test controller logic in ASP.NET Core](https://docs.microsoft.com/en-us/aspnet/core/mvc/controllers/testing?view=aspnetcore-3.0)
* Content Source: [aspnetcore/mvc/controllers/testing.md](https://github.com/aspnet/AspNetCore.Docs/blob/master/aspnetcore/mvc/controllers/testing.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-mvc**
* GitHub Login: @ardalis
* Microsoft Alias: **riande**

**Comment by Rick-Anderson** [comment]
@vanillajonathan  yes, @ardalis  is a big repository fan. If we had the resources to rewrite this, we'd do it without a repository.

The article does a good job of unit testing controllers and the repository pattern doesn't detract too much.

It would be nice to add Razor Pages too.

Most people like the article:
![image](https://user-images.githubusercontent.com/3605364/69362774-4349d900-0c33-11ea-9fe8-8afccbcc1c95.png)

**Comment by ardalis** [comment]
Thanks, Rick.

@vanillajonathan It's worth noting that the Repository pattern depends on having an abstraction that you control, and depending on that in your application, rather than on an implementation. A DbContext is an implementation. So while yes, one can say it implements Repository, for your application to properly use the pattern, your application should depend on an abstraction (an interface). This is mentioned in the tutorial you linked to as well, and is also supported by the [Dependency Inversion Principle](https://deviq.com/dependency-inversion-principle/).

I have more on the topic here if you're interested:

- https://www.weeklydevtips.com/episodes/024
- https://www.weeklydevtips.com/episodes/025

**Comment by vanillajonathan** [comment]
@Rick-Anderson shouldn't the article mention DbContext if even briefly?
To encourage it, discourage it, or mention if its possible to unit test when using a DbContext directly?
The article does a good job of explaining unit testing controllers when using the repository pattern, but if you're not using the repository pattern then the article is lacking. I can understand you guys lacking the resources to rewrite the entire article, but the DbContext could be briefly mentioned, if even just to advice against it.

@ardalis Thank you for your insightful comment, and the interesting links. I listened to the podcasts, and I found it interesting to hear your arguments.
The article on episode 24 mentions:
>  I have a link to a very useful EF Core implementation in the show notes for this episode that should help get you started or perhaps give you some ideas you can use with your existing implementation.

But I didn't find this link. Or were this referring to the repository pattern and specification pattern links on DevIQ?
